src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,889 @@
+we see that
+````math
+    \begin{align*}
+    ||\operatorname{Log}(p^+q)||_{\operatorname{Fr}}
+    &= ||\operatorname{Log}\left(
+        \operatorname{Exp}((p^{+}X)^T)
+        \operatorname{Exp}(p^{+}X - (p^{+}X)^T)
+    \right)||_{\operatorname{Fr}} \\
+    &= ||p^{+}X + \frac{1}{2}[(p^{+}X)^T, p^{+}X - (p^{+}X)^T]
+            + \ldots ||_{\operatorname{Fr}} \\
+    &≈ ||p^{+}X||_{\operatorname{Fr}} = \operatorname{dist}(p, q).
+    \end{align*}
+````
+"""
+function distance(M::Symplectic{n}, p, q) where {n}
+    return norm(log(symplectic_inverse_times(M, p, q)))
+end
+
+@doc raw"""
+    exp(M::Symplectic, p, X)
+    exp!(M::Symplectic, q, p, X)
+
+The Exponential mapping on the Symplectic manifold with the
+[`RealSymplecticMetric`](@ref) Riemannian metric.
+
+For the point ``p \in \operatorname{Sp}(2n)`` the exponential mapping along the tangent
+vector ``X \in T_p\operatorname{Sp}(2n)`` is computed as [^WangRealSymplecticGroup]
+````math
+    \operatorname{exp}_p(X) = p \operatorname{Exp}((p^{-1}X)^T)
+                                \operatorname{Exp}(p^{-1}X - (p^{-1}X)^T),
+````
+where ``\operatorname{Exp}(\cdot)`` denotes the matrix exponential.
+
+[^WangRealSymplecticGroup]:
+    > Wang, Jing and Sun, Huafei and Fiori, Simone:
+    > A Riemannian-steepest-descent approach for optimization on the real symplectic group,
+    > Mathematical Methods in the Applied Sciences, 41(11) pp. 4273-4286, 2018
+    > doi [10.1002/mma.4890](https://doi.org/10.1002/mma.4890)
+"""
+exp(::Symplectic, ::Any...)
+
+function exp!(M::Symplectic, q, p, X)
+    p_star_X = symplectic_inverse_times(M, p, X)
+    q .= p * exp(Array(p_star_X')) * exp(p_star_X - p_star_X')
+    return q
+end
+
+@doc raw"""
+    gradient(M::Symplectic, f, p, backend::RiemannianProjectionBackend;
+             extended_metric=true)
+    gradient!(M::Symplectic, f, p, backend::RiemannianProjectionBackend;
+             extended_metric=true)
+
+Compute the manifold gradient ``\text{grad}f(p)`` of a scalar function
+``f \colon \operatorname{Sp}(2n) \rightarrow ℝ`` at
+``p \in \operatorname{Sp}(2n)``.
+
+The element ``\text{grad}f(p)`` is found as the Riesz representer of the differential
+``\text{D}f(p) \colon T_p\operatorname{Sp}(2n) \rightarrow ℝ`` w.r.t.
+the Riemannian metric inner product at ``p`` [^FioriSimone2011].
+That is, ``\text{grad}f(p) \in T_p\operatorname{Sp}(2n)`` solves the relation
+````math
+    g_p(\text{grad}f(p), X) = \text{D}f(p) \quad\forall\; X \in T_p\operatorname{Sp}(2n).
+````
+
+The default behaviour is to first change the representation of the Euclidean gradient from
+the Euclidean metric to the [`RealSymplecticMetric`](@ref) at ``p``, and then we projecting
+the result onto the correct tangent tangent space ``T_p\operatorname{Sp}(2n, ℝ)``
+w.r.t the Riemannian metric ``g_p`` extended to the entire embedding space.
+
+# Arguments:
+- `extended_metric = true`: If `true`, compute the gradient ``\text{grad}f(p)`` by
```

Comment:
This should probably be generalized to our generic `gradient` but I need to think about it more.

Reply:
I think we can leave this as-is here.